pkg/storage/unified/search/document.go

```diff
@@ -0,0 +1,171 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/grafana/grafana/pkg/apimachinery/utils"
+	"github.com/grafana/grafana/pkg/services/store/kind/dashboard"
+	"github.com/grafana/grafana/pkg/storage/unified/resource"
+)
+
+type DocumentBuilderProvider interface {
+	// The list returned here defines the set of resource kinds we know about and how to
+	// convert them to documents.  Long term this will likely need to to understand
+	// the "app manifest" that would includes declarative way to identify searchable fields
+	GetDocumentBuilders(ctx context.Context) ([]resource.DocumentBuilderInfo, error)
+}
+
+// Replaced in enterprise with a version that includes stats
+func ProvideBuilders() DocumentBuilderProvider {
+	return &standardDocumentProvider{}
+}
+
+type standardDocumentProvider struct {
+	_ int // sealed
+}
+
+type defaultDocumentBuilder struct {
+	_ int // sealed
+}
+
+var (
+	_ resource.DocumentBuilder = &defaultDocumentBuilder{}
+	_ DocumentBuilderProvider  = &standardDocumentProvider{}
+)
+
+func (p *standardDocumentProvider) GetDocumentBuilders(ctx context.Context) ([]resource.DocumentBuilderInfo, error) {
+	return []resource.DocumentBuilderInfo{
+		{
+			Builder: &defaultDocumentBuilder{},
+		},
+		{
+			GroupResource: schema.GroupResource{
+				Group:    "dashboard.grafana.app",
+				Resource: "dashboards",
+			},
+
+			// This is a dummy example, and will need resolver setup for enterprise stats and and (eventually) data sources
+			Namespaced: func(ctx context.Context, namespace string, blob resource.BlobSupport) (resource.DocumentBuilder, error) {
+				lookup := dashboard.CreateDatasourceLookup([]*dashboard.DatasourceQueryResult{
+					// TODO, query data sources
+				})
+				return &DashboardDocumentBuilder{
+					Namespace:        namespace,
+					DatasourceLookup: lookup,
+					Stats:            nil, // loaded in enterprise
+					Blob:             blob,
+				}, nil
+			},
+		},
+	}, nil
+}
+
+func (*defaultDocumentBuilder) BuildDocument(_ context.Context, key *resource.ResourceKey, rv int64, value []byte) (resource.IndexableDocument, error) {
+	tmp := &unstructured.Unstructured{}
+	err := tmp.UnmarshalJSON(value)
+	if err != nil {
+		return nil, err
+	}
+
+	obj, err := utils.MetaAccessor(tmp)
+	if err != nil {
+		return nil, err
+	}
+
+	doc := &StandardDocumentFields{}
+	doc.Load(key, rv, obj)
+
+	doc.Title = obj.FindTitle(doc.Name)
+	doc.ByteSize = len(value)
+
+	return doc, nil
+}
+
+// This is common across all resources
+type StandardDocumentFields struct {
+	// unique ID across everything (group+resource+namespace+name)
+	ID string `json:"id"`
+	RV int64  `json:"rv"`
+
+	Group     string `json:"group"`
+	Resource  string `json:"resource"`
+	Namespace string `json:"namespace"`
+	Name      string `json:"name"`
+
+	Folder      string   `json:"folder,omitempty"`
+	Title       string   `json:"title,omitempty"`
+	Description string   `json:"description,omitempty"`
+	Tags        []string `json:"tags,omitempty"`
```

Comment:
everything will have tags?

Reply:
tags and title are on the "spec" I believe.  which can vary per resource right?

Reply:
however they exists across any resource, we want them in a common place in the search document.  When searching within a single document it is fine -- but "title" and "tags" should be at the same place in our index regardless where they live in the original json byte[]

Reply:
Obviously dashboards have their own expectations for what `tags` are. But for all other resources, what kind of info are `tags` meant to hold?

Reply:
> they exists across any resource
that was the question 👍